feat(max): allow editing of current dashboard

denakorita opened the pull request:
## Problem
Max cannot add insights, edit the name of a dashboard or its description. This allows that.
Editing of filters and breakdowns postponed for now due to upcoming changes in the taxonomy.


## Changes
<img width="1276" height="680" alt="Screenshot 2025-09-30 at 17 16 44" src="https://github.com/user-attachments/assets/fc0b13b9-dab8-4f68-9f5f-a598c4c7dee0" />


## How did you test this code?
Manually
Unittests and evals

denakorita committed:
add tool to edit existing dashboard

denakorita committed:
refactor

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

denakorita committed:
refactor

denakorita committed:
add the evals

denakorita committed:
Merge remote-tracking branch 'origin/master' into feat/edit-existing-dashboard

denakorita committed:
resolve mypy

kappa90 commented on ee/hogai/graph/dashboards/nodes.py:
```diff
@@ -278,22 +283,29 @@ async def _create_dashboard_with_insights(
+                    created_by=self._user,
+                )
+                if dashboard_id is None
+                else Dashboard.objects.get(id=dashboard_id)
```
This can throw an `ObjectNotFound`

kappa90 commented on ee/hogai/graph/dashboards/nodes.py:
```diff
@@ -278,22 +283,29 @@ async def _create_dashboard_with_insights(
             all_insights = list(Insight.objects.filter(id__in=insights, team=self._team))
-
+            current_insight_ids = (
+                list(dashboard.insights.all().values_list("id", flat=True)) if dashboard_id is not None else []
```
Can we prefetch the insights so we don't do two calls?

kappa90 commented on products/dashboards/backend/max_tools.py:
```diff
@@ -0,0 +1,108 @@
+    """
+
+    dashboard_name: str = Field(
+        description="The new name for the dashboard. Only provide if the user explicitly asks to rename the dashboard.",
```
It's not an optional field, so this will always be provided.

kappa90 commented on products/dashboards/backend/max_tools.py:
```diff
@@ -0,0 +1,108 @@
+    @transaction.atomic
+    def _update_dashboard_name(self, dashboard_id: int, new_name: str) -> Dashboard:
+        """Update the dashboard name."""
+        dashboard = Dashboard.objects.get(id=dashboard_id, team=self._team)
```
Can throw `ObjectNotFound`

denakorita replied:
I have a generic try except above in the `_update_dashboard_name` that captures all errors during the name update. But I see that is not good enough, I should interrupt the adding of the insights if the dashboard is not found. Will handle this.

kappa90 commented on products/dashboards/backend/max_tools.py:
```diff
@@ -0,0 +1,108 @@
+            result_message += (
+                result.messages[0].content
+                if result.messages and isinstance(result.messages[0], AssistantToolCallMessage)
+                else f"Dashboard was edited successfully. Added {len(insights_to_add)} insights to the dashboard."
```
What if the insights addition fails? This message doesn't take that into account.

denakorita replied:
If the insight addition failed then the `result.messages[0].content` should be able to handle that and give back an error message that explains that the insights could not be added and only the links to the insights that were added are mentioned

kappa90 reviewed (dismissed):
- Doesn't work on homepage
- (edge case) Doesn't let you remove insights
- Doesn't update the dashboard after it's been edited, you need to manually refresh the page
- Not your fault, but the `MaxTool` dotted line wrapper is really ugly, can we think of putting it somewhere else in the view? It also covers the three dots menu. 

<img width="1255" height="855" alt="Image" src="https://github.com/user-attachments/assets/9508455f-7a51-457d-b91b-68d889e0bb86" />

Twixes reviewed (changes requested):
Agreed with Em's points. On the broader user experience:
The tool is called "Edit dashboard", but its actual capability is so narrow, that the name is rather confusing. E.g. you can update name but not description, can add insights but not remove, can't update filters. So I see two ways out currently:
1. Split the tool into an "Add insight(s) to dashboard" one and a "Rename dashboard" one (though these are _very_ specific then)
2. Allow the "Edit dashboard" tool to: update filters, change description, remove tiles (this is the order of importance I'd say). Filters _would_ require exploring the taxonomy, but that feels like the most essential use case for editing a dashboard actually, as a user.

Also worth thinking of tools as a platform, e.g. how can we use the same paradigm for renaming a dashboard as renaming a feature flag.

denakorita committed:
add tests fix evals

denakorita committed:
small refactoring

denakorita committed:
Merge remote-tracking branch 'origin/master' into feat/edit-existing-dashboard

denakorita committed:
remove the format file

denakorita committed:
max edit should only be valid for cases where user can edit

denakorita committed:
fix margin issues, fix access control issues

denakorita committed:
add comment for the margin padding

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

denakorita committed:
fix issue with returning none

denakorita commented:
@kappa90 with regards to the home page dashboard editing, not sure if it makes sense to allow that through Max, there directly, since in the home page you cannot edit elements such as name, description or add an insight. You'd have to click "View dashboard" and then you can do so. 
I can add a button to say "edit with AI" but I am unsure of that. It might be confusing for the user, what can Max edit? Since we miss those dotted lines also.

Twixes committed:
Merge branch 'master' into feat/edit-existing-dashboard

Twixes reviewed (changes requested):
Looking good mostly! But two notes from testing this out once more:
1. When the tool completes, I'm consistently getting an error with Temporal worker logs showing `{'error': {'message': "An assistant message with 'tool_calls' 
must be followed by tool messages responding to each 'tool_call_id'. The following tool_call_ids did
not have response messages: call_a03lHBwJSdyQamcMqVpIYMlI", 'type': 'invalid_request_error', 
'param': 'messages', 'code': None}}` 
    <img width="371" height="169" alt="Image" src="https://github.com/user-attachments/assets/f2031449-6830-404a-ad40-5c2e7dba4d8c" />
2. Let's split out "Add insight(s) to dashboard" and "Edit dashboard metadata" in the end. They're massively different use cases, and the former is the one with 90% of the value – but it's not obvious that's actually available when the tool's only described as "Edit dashboard"!

denakorita committed:
Merge remote-tracking branch 'origin/master' into feat/edit-existing-dashboard

denakorita committed:
review comments add tests

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

denakorita committed:
fix issues with access control

denakorita committed:
Merge remote-tracking branch 'origin/master' into feat/edit-existing-dashboard

denakorita committed:
fix test

denakorita committed:
fix access level

denakorita committed:
revert snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

Twixes committed:
Merge branch 'master' into feat/edit-existing-dashboard

Twixes committed:
Fix pluralization of word

Adding a universal util `pluralize()` which we've had in TypeScript, but not Python!

Twixes committed:
Fix tests after merge (post loop revamp)

denakorita committed:
fix tests

Twixes reviewed (approved):
The "Add insight" framing makes this more powerful! Let's go.

It's not quite perfectly reliable, as we're missing one bit: the LLM _can_ choose something that looks like it could be useful, when it's not. See this case:

<img width="553" height="290" alt="Screenshot 2025-10-14 at 19 14 18" src="https://github.com/user-attachments/assets/d6824274-38cd-4c6a-9128-486f4a9bec18" />

If we can smooth it out with more context provided to the LLM to review its own work, that'd be great. But already worth getting in as is.

denakorita committed:
Merge remote-tracking branch 'origin/master' into feat/edit-existing-dashboard

denakorita committed:
refix tests

denakorita committed:
refix tests

denakorita committed:
Merge branch 'master' into feat/edit-existing-dashboard

denakorita committed:
Merge branch 'master' into feat/edit-existing-dashboard

denakorita commented:
@Twixes There's a few flaky tests here failing, not related to my code changes though, I would like to merge this if you have no objections

denakorita committed:
Merge branch 'master' into feat/edit-existing-dashboard

Twixes commented:
Let's merge.
